src/disc/stk/Albany_GenericSTKMeshStruct.cpp

```diff
@@ -1393,21 +1398,119 @@ void Albany::GenericSTKMeshStruct::loadField (const std::string& field_name, con
+    TEUCHOS_TEST_FOR_EXCEPTION(layered, std::logic_error, "Error! Layered fields cannot be computed from a mathematical expression.\n");
+
+    int field_dim = 1;
+    if (!scalar) {
+      TEUCHOS_TEST_FOR_EXCEPTION(!params.isParameter("Vector Length"), std::logic_error,
+                                 "Error! In order to compute a vector field from a mathematical expression, "
+                                 "you must provide the parameter 'Vector Length'.\n");
+      field_dim = params.get<int>("Vector Length");
+    }
+
+    // Get the expressions out of the parameter list.
+    Teuchos::Array<std::string> expressions = params.get<Teuchos::Array<std::string>>("Field Expression");
+
+    // NOTE: we need expressions to be of length AT LEAST equal to the field dimension.
+    //       If the length L is larger than the field dimension M, then the first L-M
+    //       strings are assumed to be coefficients needed for the field formula.
+    //       E.g.: if we have a field of dimension 2, one could write
+    //         <Parameter name="Field Expression" type="Array(string)" value="{a=1.5;b=-1;c=2;a*x^2+b*x+c;a*x+b*x+c}"/>
+
+    int num_expr = expressions.size();
+    TEUCHOS_TEST_FOR_EXCEPTION(num_expr<field_dim, Teuchos::Exceptions::InvalidParameter,
+                               "Error! Input array for 'Field Expression' is too short. "
+                               "Expected length >=" << field_dim << ". Got " << num_expr << " instead.\n");
+    panzer::Expr::Eval<double*,Kokkos::HostSpace> eval;
+    Teuchos::any result;
+
+    // Start by reading the parameters used in the field expression(s)
+    int num_params = num_expr - field_dim;
+    std::string expr_params;
+    for (int iparam=0; iparam<num_params; ++iparam) {
+      expr_params += expressions[iparam] + ";";
+    }
+
+    *out << "  - Computing " << field_type << " field '" << field_name << "' from mathematical expression(s):";
+    for (int idim=num_params; idim<num_expr; ++idim) {
+      *out << " " << expressions[idim] << (idim==num_expr-1 ? "" : ";");
+    }
+    if (num_params>0) {
+      *out << " (with";
+      for (int idim=0; idim<num_params; ++idim) {
+        *out << " " << expressions[idim] << (idim==num_params-1 ? "" : ";");
+      }
+      *out << ")";
+    }
+    *out << " ... ";
+
+    // Now loop through the nodes, and read the components
+    const auto& coordinates = *this->getCoordinatesField3d();
+    stk::mesh::EntityId gid;
+    LO lid;
+    double* values;
+    double* xyz;
+    stk::topology::rank_t entity_rank = stk::topology::NODE_RANK;
+    if (scalar) {
+      AbstractSTKFieldContainer::ScalarFieldType* field = metaData->get_field<AbstractSTKFieldContainer::ScalarFieldType> (entity_rank, field_name);
+      for (int i(0); i<entities.size(); ++i)
+      {
+        values = stk::mesh::field_data(*field, entities[i]);
+        xyz    = stk::mesh::field_data(coordinates, entities[i]);
+
+        eval.set("x",xyz[0]);
```

Comment:
This is a slow way to process the data. It would be better to create a Kokkos::View for all the X values for all entities and then evaluate the expression for all entities and get the resulting Kokkos::View. that is how the system was designed to be used.

Reply:
I agree, but I couldn't find a way to extract all the xyz values at once. The only way to do it would be to loop, fill the xyz view, then call the eval method once. I guess that can be done. I didn't bother with that since this is a setup phase only, but I guess for complex math expressions it may be worth it.

I'll fix this.

Reply:
It should be possible to get array data out of STK somehow... might be very complicated though, maybe the fill loop is the easier approach, I don't know.